src/cmap/connection_pool.ts

```diff
@@ -463,26 +485,29 @@ function processWaitQueue(pool: any) {
     }
 
     const connection = pool[kConnections].shift();
-    const isStale = connectionIsStale(pool, connection);
-    const isIdle = connectionIsIdle(pool, connection);
-    if (!isStale && !isIdle && !connection.closed) {
-      pool.emit('connectionCheckedOut', new ConnectionCheckedOutEvent(pool, connection));
-      clearTimeout(waitQueueMember.timer);
-      pool[kWaitQueue].shift();
-      waitQueueMember.callback(undefined, connection);
-      return;
+    if (connection) {
+      const isStale = connectionIsStale(pool, connection);
+      const isIdle = connectionIsIdle(pool, connection);
+      if (!isStale && !isIdle && !connection.closed) {
+        pool.emit('connectionCheckedOut', new ConnectionCheckedOutEvent(pool, connection));
+        if (waitQueueMember && waitQueueMember.timer) {
```

Comment:
This is a bug, you should only be gating the call to `clearTimeout` on this:
```
if (waitQueueMember && waitQueueMember.timer) {
  clearTimeout(waitQueueMember.timer);
}

pool[kWaitQueue].shift();
waitQueueMember.callback(undefined, connection);
```

Additionally, I think both with `connection` and `waitQueueMember` we should take a "bail-early" strategy:
```
while (pool.waitQueueSize) {
  const waitQueueMember = pool[kWaitQueue].peekFront();
  if (waitQueueMember == null) {
    throw new Error('should never happen!');
  }
  
  ....
  const connection = pool[kConnections].shift();
  if (connection == null) {
    throw new Error('should never happen!');
  }
  ...
```
I'm throwing errors here for effect, we should probably not do that. Related question: these checks are redundant, at least in one case:
```
while (pool.waitQueueSize) {
  const waitQueueMember = pool[kWaitQueue].peekFront();
```
that guarantees that we always have something on the wait queue before peeking, and we're only peeking because we might want to leave it in place. Perhaps there are some cases where we would be comfortable using the bang (`!`).

Reply:
Nice find, thanks for taking a look, I agree it is redundant, if we do start using the assert `!` maybe we should enable the linter to warn on them so we have to comment an eslint-disable line to ignore it? or is that tooling overkill?